File: apps/api/src/api/modules/stream/playlist.py
"""Rewriting the URLs inside an `.m3u8`, so nested requests come back here.

Without this the endpoint is pointless for HLS. A player would read the master
playlist through us, then go and fetch the variants and segments straight from
the origin — which is exactly the request a browser refuses, and the reason
anybody proxies a playlist in the first place.

The links written back are **relative** (`/stream?url=…`), so this does not care
what address it is reached on: localhost, a tunnel, or a hostname it will be
given next year.
"""

import re
from urllib.parse import quote, urljoin

#: `#EXT-X-KEY`, `#EXT-X-MAP` and `#EXT-X-MEDIA` all hide a URL in an attribute
#: rather than on a line of their own.
_URI_ATTR = re.compile(r'URI="([^"]*)"')

#: Where a rewritten link points. One place, so the route and the rewriter
#: cannot drift apart.
PATH = "/stream"


def is_playlist(content_type: str | None, url: str) -> bool:
    """Whether this response is a playlist rather than media.

    The content type first, because it is what the origin says it is sending;
    the extension second, because plenty of them send `application/octet-stream`
    and mean `.m3u8`.
    """
    if content_type and "mpegurl" in content_type.lower():
        return True
    return url.split("?", 1)[0].lower().endswith(".m3u8")


def proxied(url: str) -> str:
    return f"{PATH}?url={quote(url, safe='')}"


def rewrite(text: str, base_url: str) -> str:
    """Every URL in [text] pointed back here, resolved against [base_url].

    `base_url` is the address the playlist was *finally* fetched from — after
    redirects — because that is what its relative lines are relative to.
    """
    lines = []

    for line in text.splitlines():
        stripped = line.strip()
        if not stripped:
            lines.append(line)
        elif stripped.startswith("#"):
            lines.append(
                _URI_ATTR.sub(lambda m: f'URI="{proxied(urljoin(base_url, m.group(1)))}"', line)
            )
        else:
            lines.append(proxied(urljoin(base_url, stripped)))

    return "\n".join(lines) + "\n"

File: packages/dart/cors_proxy/lib/src/playlist.dart
import 'proxy_url.dart';

/// Whether a response is an HLS playlist, and so worth reading before passing
/// on.
///
/// Content type first, because it is the thing that is actually declared; the
/// extension is the fallback for a CDN that serves `application/octet-stream`
/// and lets the player work it out.
bool looksLikePlaylist({String? contentType, required String path}) {
  final type = contentType?.toLowerCase() ?? '';
  if (type.contains('mpegurl')) return true;
  return path.toLowerCase().endsWith('.m3u8');
}

/// A playlist with every address in it pointed back at the proxy.
///
/// A relative name needs nothing done to it — the proxied address keeps the
/// upstream's directory structure, so `seg-1.ts` resolves correctly on its own.
/// An absolute one does: ashdi's master playlist names each variant in full,
/// and a player following those would leave this origin and hit exactly the
/// wall the proxy exists to get around. That is not a hypothetical — it is what
/// ashdi actually sends.
String rewritePlaylist(String body, {String base = ''}) {
  final lines = body.split('\n');

  for (var i = 0; i < lines.length; i++) {
    final line = lines[i];
    final trimmed = line.trim();
    if (trimmed.isEmpty) continue;

    if (trimmed.startsWith('#')) {
      // A key or an initialisation segment is named inside the tag rather than
      // on a line of its own, and an encrypted stream that cannot fetch its key
      // plays nothing at all.
      lines[i] = _rewriteUris(line, base);
      continue;
    }

    lines[i] = _rewrite(trimmed, base) ?? line;
  }

  return lines.join('\n');
}

/// `URI="..."` wherever it appears in a tag.
String _rewriteUris(String line, String base) =>
    line.replaceAllMapped(RegExp(r'URI="([^"]*)"'), (match) {
      final rewritten = _rewrite(match[1]!, base);
      return rewritten == null ? match[0]! : 'URI="$rewritten"';
    });

/// The proxied form of an absolute address, or `null` when there is nothing to
/// do — which is the common case and has to stay cheap.
String? _rewrite(String value, String base) {
  if (!value.startsWith('http://') && !value.startsWith('https://')) {
    return null;
  }
  final url = Uri.tryParse(value);
  if (url == null || url.host.isEmpty) return null;
  return ProxyUrl.encode(url, base: base);
}
